src/common/include/gudhi/uint128.h

```diff
@@ -0,0 +1,147 @@
+    Fake_uint128 res;
+    res.low = a.low & b.low;
+    res.high = a.high & b.high;
+    GUDHI_VERIF (a.native() & b.native());
+    return res;
+  }
+  friend Fake_uint128 operator|(Fake_uint128 a, Fake_uint128 b){
+    Fake_uint128 res;
+    res.low = a.low | b.low;
+    res.high = a.high | b.high;
+    GUDHI_VERIF (a.native() | b.native());
+    return res;
+  }
+  friend Fake_uint128 operator~(Fake_uint128 a){
+    Fake_uint128 res;
+    res.low = ~a.low;
+    res.high = ~a.high;
+    return res;
+  }
+  // In-place arithmetic
+  Fake_uint128& operator+=(Fake_uint128 a) { *this = *this + a; return *this; }
+  Fake_uint128& operator-=(Fake_uint128 a) { *this = *this - a; return *this; }
+  Fake_uint128& operator++() { if (++low == 0) ++high; return *this; }
+  Fake_uint128& operator--() { if (low-- == 0) --high; return *this; }
+  Fake_uint128& operator<<=(uint8_t a) { *this = *this << a; return *this; }
+  Fake_uint128& operator>>=(uint8_t a) { *this = *this >> a; return *this; }
+  Fake_uint128& operator&=(Fake_uint128 a) { *this = *this & a; return *this; }
+  Fake_uint128& operator|=(Fake_uint128 a) { *this = *this | a; return *this; }
+  // Comparisons
+  friend bool operator==(Fake_uint128 a, Fake_uint128 b){
+    return a.low == b.low && a.high == b.high;
+  }
+  friend bool operator!=(Fake_uint128 a, Fake_uint128 b){
+    return a.low != b.low || a.high != b.high;
+  }
+  friend bool operator<(Fake_uint128 a, Fake_uint128 b){
+    return a.high < b.high || (a.high == b.high && a.low < b.low);
+  }
+  friend bool operator>(Fake_uint128 a, Fake_uint128 b){
+    return a.high > b.high || (a.high == b.high && a.low > b.low);
+  }
+  friend bool operator<=(Fake_uint128 a, Fake_uint128 b){
+    return a.high < b.high || (a.high == b.high && a.low <= b.low);
+  }
+  friend bool operator>=(Fake_uint128 a, Fake_uint128 b){
+    return a.high > b.high || (a.high == b.high && a.low >= b.low);
+  }
+  // Misc
+  friend std::size_t hash_value(Fake_uint128 a) {
+    typedef std::pair<std::uint64_t, std::uint64_t> P;
+    return boost::hash_value(P(a.high, a.low));
+  }
+  template <class T, class=std::enable_if_t<std::is_integral_v<T>>>
+  explicit operator T() const {
+    GUDHI_CHECK(high == 0 && low <= std::numeric_limits<T>::max(), "");
+    return static_cast<T>(low);
+  }
+  private:
+  std::uint64_t high, low; // does the order matter?
+ #undef GUDHI_VERIF
+};
+typedef Fake_uint128 uint128_t;
+} // namespace Gudhi::numbers
+template<> class std::numeric_limits<Gudhi::numbers::Fake_uint128> {
+  public:
+  static constexpr bool is_specialized = true;
+  static constexpr bool is_signed = false;
+  static constexpr bool is_integer = true;
+  static constexpr bool is_exact = true;
+  static constexpr bool has_infinity = false;
+  static constexpr bool is_modulo = true;
+  static constexpr int digits = 128;
+  static constexpr int radix = 2;
+  // etc
+};
+#else
+namespace Gudhi::numbers {
+typedef unsigned __int128 uint128_t;
```

Comment:
ISTR that it was the first thing I tried but it was significantly worse, but I'll check again.
The current code is supposed to be compatible with 32 bit CPUs: `long long` has to be at least 64 bits in C++ (possibly through emulation).

Reply:
I checked again. On linux, `boost::multiprecision::uint128_t` delegates to `unsigned __int128` for storage and most operations. On some testcases, it has a bit of overhead (10%, mostly because of checks in `operator<<` and `operator>>`). On windows, there is no 128bit type, and boost implements it as a bignum, so it takes 50% more storage (it stores the size), and operations are generally slower. On some testcases it is not noticable, but I have a testcase (a bit pathological: asking for homology in dim 0-25 for a circle with 24 points) where it multiplies the running time by more than 4 compared to linux. Fake_uint128 has less than 10% overhead on the same testcase. Saying that we don't care about windows is tempting, but if we do care, a factor 4 in the running time seems worth it.

Reply:
Ok, let's keep your implementation in this case, but if you could add some comments to explain why we don't go with `boost::multiprecision::uint128_t`, that would be appreciate, and we can remove it later if boost is doing some imrpovements on windows side.